sdk/pinocchio/src/sysvars/slot_hashes.rs

```diff
@@ -0,0 +1,895 @@
+//! Efficient, zero-copy access to SlotHashes sysvar data.
+
+use crate::{
+    account_info::{AccountInfo, Ref},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+    sysvars::clock::Slot,
+};
+use core::{mem, ops::Deref, ptr};
+
+/// SysvarS1otHashes111111111111111111111111111
+pub const SLOTHASHES_ID: Pubkey = [
+    6, 167, 213, 23, 25, 47, 10, 175, 198, 242, 101, 227, 251, 119, 204, 122, 218, 130, 197, 41,
+    208, 190, 59, 19, 110, 45, 0, 85, 32, 0, 0, 0,
+];
+pub const MAX_ENTRIES: usize = 512;
+pub const HASH_BYTES: usize = 32;
+
+/// Sysvar data is:
+/// len    (8 bytes): little-endian entry count (≤ 512)
+/// entries(len × 40 bytes):    consecutive `(u64 slot, [u8;32] hash)` pairs
+pub const NUM_ENTRIES_SIZE: usize = mem::size_of::<u64>();
+pub const SLOT_SIZE: usize = mem::size_of::<Slot>();
+pub const ENTRY_SIZE: usize = SLOT_SIZE + HASH_BYTES;
+
+/// A single entry in the `SlotHashes` sysvar.
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[repr(C)]
+pub struct SlotHashEntry {
+    /// The slot number.
+    pub slot: Slot,
+    /// The hash corresponding to the slot.
+    pub hash: [u8; HASH_BYTES],
+}
+
+/// Provides zero-copy access to the data of the `SlotHashes` sysvar.
+///
+/// Internally it keeps either a plain slice `&'a [u8]` *or* the `Ref<'a, [u8]>`
+/// returned by `AccountInfo::try_borrow_data()`.  Holding the `Ref` variant is
+/// important because dropping the `Ref` would release the runtime borrow while
+/// users still hold `&[u8]` references obtained from the struct.
+enum Data<'a> {
+    Slice(&'a [u8]),
+    Ref(Ref<'a, [u8]>),
+}
+
+impl core::ops::Deref for Data<'_> {
+    type Target = [u8];
+    #[inline(always)]
+    fn deref(&self) -> &Self::Target {
+        match self {
+            Data::Slice(s) => s,
+            Data::Ref(r) => r,
+        }
+    }
+}
+
+/// SlotHashes provides read-only, zero-copy access to SlotHashes sysvar bytes.
+pub struct SlotHashes<'a> {
+    data: Data<'a>,
```

Comment:
Not sure if you considered using a generic type instead of the `enum` – e.g., `T: Deref<Target = [u8]>`. This would avoid creating a type just to hold the reference.

Reply:
I agree. The enum was meant to ensure that the `Ref<'a, [u8]>` returned by `AccountInfo::try_borrow_data()` is kept alive for the whole lifetime of the view. A plain `T: Deref` erases the fact that one variant is a runtime borrow needing to be held.

But this isn't a concern when used by Solana programs, only when used in other contexts.

I considered wrapping the Ref to collapse the API to a single type while still preventing the Ref from being dropped prematurely, but your solution is cleaner, given that pinocchio [expressly asks devs not to use it in other contexts](https://github.com/anza-xyz/pinocchio/blob/887a1150bba40c5c77373da90b76557d8c73b747/sdk/pinocchio/src/lib.rs#L9-L10). Will implement.

Reply:
Addressed in dce5e60da08b583d82c3eee70c6114ba52a2bc8b